nomos-sdp/src/state.rs

```diff
@@ -0,0 +1,615 @@
+        }
+    }
+
+    const fn try_into_withdrawn<ContractAddress>(
+        mut self,
+        block_number: BlockNumber,
+        event_type: EventType,
+        service_params: &ServiceParameters<ContractAddress>,
+    ) -> Result<WithdrawnState, InactiveStateError> {
+        match event_type {
+            EventType::Withdrawal => {
+                if self.0.created.wrapping_add(service_params.lock_period) >= block_number {
+                    return Err(InactiveStateError::ToWithdrawalWhileLocked);
+                }
+                self.0.withdrawn = Some(block_number);
+                Ok(WithdrawnState(self.0))
+            }
+            EventType::Declaration | EventType::Reward => {
+                Err(InactiveStateError::ToWithdrawalInvalidEvent(event_type))
+            }
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum WithdrawnStateError {
+    #[error("Withdawn state can be rewarded only during the same blcok")]
+    ToWithdrawnRewardedNotSameBlock,
+    #[error("Withdrawn can not transition to withdrawn rewarded during {0:?} event")]
+    ToWithdrawnRewardedInvalidEvent(EventType),
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Hash)]
+pub struct WithdrawnState(ProviderInfo);
+
+impl WithdrawnState {
+    fn try_into_withdrawn_rewarded(
+        mut self,
+        block_number: BlockNumber,
+        event_type: EventType,
+    ) -> Result<Self, WithdrawnStateError> {
+        if matches!(event_type, EventType::Reward) {
+            // Only allow reward for withdrawal in the same block number.
+            if self.0.withdrawn != Some(block_number) {
+                return Err(WithdrawnStateError::ToWithdrawnRewardedNotSameBlock);
+            }
+
+            self.0.rewarded = Some(block_number);
+            Ok(self)
+        } else {
+            Err(WithdrawnStateError::ToWithdrawnRewardedInvalidEvent(
+                event_type,
+            ))
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ProviderStateError {
+    #[error(transparent)]
+    Active(#[from] ActiveStateError),
+    #[error(transparent)]
+    Inactive(#[from] InactiveStateError),
+    #[error(transparent)]
+    Withdrawn(#[from] WithdrawnStateError),
+    #[error("Withdrawn can not transition to active")]
+    WithdrawnToActive,
+    #[error("Provided block is in past, can not transition to state in previous block")]
+    BlockFromPast,
+}
+
+#[derive(Debug, Eq, PartialEq, Hash)]
+pub enum ProviderState {
+    Active(ActiveState),
+    Inactive(InactiveState),
+    Withdrawn(WithdrawnState),
+}
+
+#[expect(
+    clippy::from_over_into,
+    reason = "ProviderState can not be derived from ProviderInfo alone"
+)]
+impl Into<ProviderInfo> for ProviderState {
+    fn into(self) -> ProviderInfo {
+        match self {
+            Self::Active(active_state) => active_state.0,
+            Self::Inactive(inactive_state) => inactive_state.0,
+            Self::Withdrawn(withdrawn_state) => withdrawn_state.0,
+        }
+    }
+}
```

Comment:
Wait, why this cannot be a `from` impl? 
```rust
impl From<ProviderState> for ProviderInfo {
  fn from(state: :ProviderState) -> ProviderInfo {
    match state {
            ProviderState::Active(active_state) => active_state.0,
            ProviderState::Inactive(inactive_state) => inactive_state.0,
            ProviderState::Withdrawn(withdrawn_state) => withdrawn_state.0,
        }
}
}
```

Reply:
Added a clippy comment above: `ProviderState can not be derived from ProviderInfo alone` - we need `ServiceParams` and `BlockNumber` to derive `ProviderState` from `ProviderInfo`, but we can directly derive `ProviderState` into `ProviderInfo`.

Reply:
I do not see the difference. Where are the `ServiceParams` and `BlockNumber` in the current impl?